prov/efa/src/efa_fabric.c

```diff
@@ -479,11 +491,39 @@ static int efa_get_gdr_support(char *device_name)
 		goto out;
 	}
 
-	ret = fi_read_file(get_sysfs_path(), gdr_path, &gdr_support, sizeof(char));
-	if (ret < 0)
+	ret = fi_read_file(get_sysfs_path(), gdr_path,
+			   gdr_support_str, MAX_GDR_SUPPORT_STRLEN-1);
+	if (ret < 0) {
+		if (errno == ENOENT) {
+			/* sysfs file does not exist, gdr is not supported */
+			ret = 0;
+		}
+
 		goto out;
+	}
+
+	if (ret == 0) {
+		EFA_WARN(FI_LOG_FABRIC, "Sysfs file %s is empty\n", gdr_path);
+		ret = -FI_EINVAL;
+		goto out;
+	}
+
+	gdr_support_str[ret] = '\0';
```

Comment:
This is not safe.  If the string is 80 characters or more long and does not end in a \n, then ret will be 80.  This will write into the 81st byte.

Reply:
When we call ```fi_read_file```, we used ```MAX_GDR_SUPPORT_STRLEN - 1```, not ```MAX_GDR_SUPPORT_STRLEN```. so ```ret``` will always be <= ```MAX_GDR_SUPPORT_STRLEN-1```, so is safe to use here.

Reply:
ugh; please don't do that.  It's very unclear what you're doing, and that doesn't fit with normal usage.

Reply:
I see, thanks for pointing that out! In new revision, I used MAX_GDR_SUPPORT_STRLEN to call ```fi_read_file``` and used ```gdr_support_str[MAX_GDR_SUPPORT_STRLEN-1] = '\0';```
which should always be safe.